driver-core/src/test/java/com/datastax/driver/core/CCMBridge.java

```diff
@@ -792,6 +807,81 @@ public void setWorkload(int node, Workload... workload) {
     execute(CCM_COMMAND + " node%d setworkload %s", node, workloadStr);
   }
 
+  private static VersionNumber getScyllaVersionThroughCcm(String versionString) {
+    try {
+      executeStatic("ccm create get_version -n 1 --scylla --version %s", versionString);
+      String versionOutput = executeStatic("ccm node1 versionfrombuild");
+      executeStatic("ccm remove");
+      return VersionNumber.parse(versionOutput.trim());
+    } catch (RuntimeException cause) {
+      throw new RuntimeException(
+          "Error during getting Scylla version through ccm commands.", cause);
+    }
+  }
+
+  // Not recommended for general use. Static version of execute() with minor modifications.
+  private static String executeStatic(String command, Object... args) {
```

Comment:
Can you please make static method so that instance method could use it, instead of duplicating the code.

Reply:
The problem is that for some reason i cannot.
Previous implementation looked mostly like this:
```
private static String executeStatic(File ccmDir, String command, Object... args) {
    String fullCommand;
    if (ccmDir != null) fullCommand = String.format(command, args) + " --config-dir=" + ccmDir;
    else fullCommand = String.format(command, args);
    Closer closer = Closer.create();
    // 10 minutes timeout
    ExecuteWatchdog watchDog = new ExecuteWatchdog(TimeUnit.MINUTES.toMillis(10));
    StringWriter sw = new StringWriter();
    final PrintWriter pw = new PrintWriter(sw);
    closer.register(pw);
    try {
      logger.info("Executing: " + fullCommand);
      CommandLine cli = CommandLine.parse(fullCommand);
      Executor executor = new DefaultExecutor();
      LogOutputStream outStream =
          new LogOutputStream() {
            @Override
            protected void processLine(String line, int logLevel) {
              String out = "ccmout> " + line;
              logger.debug(out);
              pw.println(out);
            }
          };
      LogOutputStream errStream =
          new LogOutputStream() {
            @Override
            protected void processLine(String line, int logLevel) {
              String err = "ccmerr> " + line;
              logger.error(err);
              pw.println(err);
            }
          };
      closer.register(outStream);
      closer.register(errStream);
      ExecuteStreamHandler streamHandler = new PumpStreamHandler(outStream, errStream);
      executor.setStreamHandler(streamHandler);
      executor.setWatchdog(watchDog);
      int retValue = executor.execute(cli, ENVIRONMENT_MAP);
      if (retValue != 0) {
        logger.error(
            "Non-zero exit code ({}) returned from executing ccm command: {}",
            retValue,
            fullCommand);
        pw.flush();
        throw new CCMException(
            String.format(
                "Non-zero exit code (%s) returned from executing ccm command: %s",
                retValue, fullCommand),
            sw.toString());
      }
    } catch (IOException e) {
      if (watchDog.killedProcess())
        logger.error("The command {} was killed after 10 minutes", fullCommand);
      pw.flush();
      throw new CCMException(
          String.format("The command %s failed to execute", fullCommand), sw.toString(), e);
    } finally {
      try {
        closer.close();
      } catch (IOException e) {
        Throwables.propagate(e);
      }
    }
    return sw.toString();
  }

  private String execute(String command, Object... args) {
    return executeStatic(this.ccmDir, command, args);
  }
```
Same as former `execute` except first two lines. What ends up happening is that when creation of `get_version` cluster returns an error, this `logger.error(` line never outputs and debugger loses track of everything afterwards. It can be replaced with any other method (like `logger.hashcode();`) and it will still hang. Removing logger resolves the problem.
```
protected void processLine(String line, int logLevel) {
              String err = "ccmerr> " + line;
              logger.error(err);
              pw.println(err);
            }
```
`jstack -F` says that every thread is in a blocked state. Not sure what is going on there.

Reply:
it is because it fails into `    } catch (IOException e) {` and never get logged, fix it by adding:
```
    } catch (IOException e) {
      if (watchDog.killedProcess())
        logger.error("The command ({}) was killed after 10 minutes with error {}", fullCommand, e.getMessage());
      else {
        logger.error("The command ({}) failed to execute with error {}", fullCommand, e.getMessage());
      }
      pw.flush();
      throw new CCMException(
          String.format("The command %s failed to execute", fullCommand), sw.toString(), e);
    } finally {
```
And it will work properly.

Reply:
Another solution is that you don't assert:
```
    } finally {
      assert bridgeA != null;
      bridgeA.close();
    }
```
And just don't close it is the case:
```
    } finally {
      if (bridgeA != null) {
        bridgeA.close();
      }
    }
```

Reply:
> it is because it fails into ` } catch (IOException e) {` and never get logged, fix it by adding:
> [...]
> And it will work properly.

I don't think that's it, but anyway I tried it and it did not help.



> Another solution is that you don't assert:
> 
> ```
>     } finally {
>       assert bridgeA != null;
>       bridgeA.close();
>     }
> ```
> 
> And just don't close it is the case:
> 
> ```
>     } finally {
>       if (bridgeA != null) {
>         bridgeA.close();
>       }
>     }
> ```

The code hangs during first (modified as above) `executeStatic()`

Reply:
hungs ? or throws exception ? where exactly it hangs ? what causes it ?

Reply:
Same as before, hangs on `logger.error` in this part:
```
LogOutputStream errStream =
          new LogOutputStream() {
            @Override
            protected void processLine(String line, int logLevel) {
              String err = "ccmerr> " + line;
              logger.error(err);
              pw.println(err);
            }
          };
```
One of the threads handling one of the output streams just dies afterwards. I don't know what is the cause.

Reply:
Actually I think it's the case that one of the streams is successfully closed. The one trying to output the error hangs and blocks the executor shutdown because it's still alive somehow.

Reply:
> Same as before, hangs on `logger.error` in this part:
> 
> ```
> LogOutputStream errStream =
>           new LogOutputStream() {
>             @Override
>             protected void processLine(String line, int logLevel) {
>               String err = "ccmerr> " + line;
>               logger.error(err);
>               pw.println(err);
>             }
>           };
> ```
> 
> One of the threads handling one of the output streams just dies afterwards. I don't know what is the cause.

Can you please share how to reproduce it ?

Reply:
1. Checkout the code from this PR
2. Apply attached patch
3. Run AsyncQueryTest with VM options: `-ea -Dscylla.version=unstable/master:latest`
[repro_patch.txt](https://github.com/user-attachments/files/18082099/repro_patch.txt)

Reply:
Seems like there is some strange race condition. just add `Logger logger` to `executeStatic` and it is going to go away.

Reply:
But it does not invalidates what i said before, if you don't drop assert or start logging an error, when CCM fails in `getVersion` you are going to lose it and it will be hard to say what exactly failed.

Reply:
> Seems like there is some strange race condition. just add `Logger logger` to `executeStatic` and it is going to go away.

That indeed works, but in that case i would still opt for having `executeStatic` as a separate method to use during static initialization and leave `execute` as it is and its usages in all of the tests. I don't know how costly it would be to instantiate a new logger for each ccm command during testing, but we probably don't want that.

> But it does not invalidates what i said before, if you don't drop assert or start logging an error, when CCM fails in getVersion you are going to lose it and it will be hard to say what exactly failed.

I'm not sure I follow this part of the problem. `assert bridgeA != null;` is a part of the test unrelated to this change. Original `execute` already logs the errors and similarly `executeStatic` does (although currently through `System.err.println` instead of a logger)

Anyway I can make the change to the `executeStatic` to use a logger instantiated on call to the method, but I feel unsure about that. While it works for me locally, it still will be called for static initialization and if the root cause is linked to initialization of the logging library this may resurface eventually.

Reply:
> > Seems like there is some strange race condition. just add `Logger logger` to `executeStatic` and it is going to go away.
> 
> That indeed works, but in that case i would still opt for having `executeStatic` as a separate method to use during static initialization and leave `execute` as it is and its usages in all of the tests. I don't know how costly it would be to instantiate a new logger for each ccm command during testing, but we probably don't want that.

Why would you `instantiate a new logger` ? Just pick static logger from the class, it works just fine.

> > But it does not invalidates what i said before, if you don't drop assert or start logging an error, when CCM fails in getVersion you are going to lose it and it will be hard to say what exactly failed.
> 
> I'm not sure I follow this part of the problem. `assert bridgeA != null;` is a part of the test unrelated to this change. Original `execute` already logs the errors and similarly `executeStatic` does (although currently through `System.err.println` instead of a logger)

It is kind of unreleated, but it leads to bad UX (in terms of error debugging) in the case that is created by this PR.
We need to fix it in the same PR, so that reviewer could have a reason why it was done.

> Anyway I can make the change to the `executeStatic` to use a logger instantiated on call to the method, but I feel unsure about that. While it works for me locally, it still will be called for static initialization and if the root cause is linked to initialization of the logging library this may resurface eventually.

If it is resurface we can fix it when it happens.

Reply:
> > > Seems like there is some strange race condition. just add `Logger logger` to `executeStatic` and it is going to go away.
> > 
> > 
> > That indeed works, but in that case i would still opt for having `executeStatic` as a separate method to use during static initialization and leave `execute` as it is and its usages in all of the tests. I don't know how costly it would be to instantiate a new logger for each ccm command during testing, but we probably don't want that.
> 
> Why would you `instantiate a new logger` ? Just pick static logger from the class, it works just fine.

The reproducer that hangs uses the static logger from the class. I guess then you meant that I should add `Logger logger = CCMBridge.logger;` to the method. That for some weird reason helps too, even though in the end it's the same logger.

> > > But it does not invalidates what i said before, if you don't drop assert or start logging an error, when CCM fails in getVersion you are going to lose it and it will be hard to say what exactly failed.
> > 
> > 
> > I'm not sure I follow this part of the problem. `assert bridgeA != null;` is a part of the test unrelated to this change. Original `execute` already logs the errors and similarly `executeStatic` does (although currently through `System.err.println` instead of a logger)
> 
> It is kind of unreleated, but it leads to bad UX (in terms of error debugging) in the case that is created by this PR. We need to fix it in the same PR, so that reviewer could have a reason why it was done.

I don't understand what is the new case that is introduced in this PR. The logging will be the same as before, except the execute method will be static.

Reply:
Ok, If i understand this right now, you're just saying that I should update the `DnsEndpointTests` at the same time in this PR by removing/changing these asserts in it. 
This does not affect the hanging at all so I was confused.

Reply:
Yes, can you do that please, it doesn ot affect hanging, but if affects what you see when ccm throw you an exception.
Currently assert exception overrides original exception, so you left with nothing.